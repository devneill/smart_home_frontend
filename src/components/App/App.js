import logo from "../../logo.png";
import "./App.style.css";
import DeviceList from "../DeviceList/DeviceList";
import Button from "../Button/Button";
import { useHistory } from "react-router-dom";

function App() {
  let history = useHistory();

  const handleClick = () => {
    history.push(`/device/new`);
  };

  return (
    <div className="App">
      <header className="App-header">
        <img src={logo} className="App-logo" alt="logo" />
      </header>
      <DeviceList />
      <Button text="Add device" onClick={handleClick} />
    </div>
  );
}

export default App;
