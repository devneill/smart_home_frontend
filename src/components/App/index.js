import { useState, useEffect } from "react";
import { BrowserRouter, Route, Switch } from "react-router-dom";
import Home from "../../pages/Home";
import DeviceDetails from "../../pages/DeviceDetails";
import DeviceCreate from "../../pages/DeviceCreate";
import NotFound from "../../pages/NotFound";

function App() {
  const [error, setError] = useState(null);
  const [isLoaded, setIsLoaded] = useState(false);
  const [devices, setDevices] = useState([]);

  useEffect(() => {
    fetch("https://canary-homework-test.herokuapp.com/devices")
      .then((res) => res.json())
      .then(
        (result) => {
          setIsLoaded(true);
          setDevices(result);
        },
        (error) => {
          setIsLoaded(true);
          setError(error);
        }
      );
  }, []);

  return (
    <BrowserRouter>
      <Switch>
        <Route
          exact
          path="/"
          render={() => (
            <Home
              error={error}
              isLoaded={isLoaded}
              devices={devices}
              setDevices={setDevices}
            />
          )}
        />
        <Route
          path="/device/new"
          render={() => (
            <DeviceCreate devices={devices} setDevices={setDevices} />
          )}
        />
        <Route
          path="/device/:deviceId"
          render={({ match }) => (
            <DeviceDetails match={match} devices={devices} />
          )}
        />
        <Route component={NotFound} />
      </Switch>
    </BrowserRouter>
  );
}

export default App;
