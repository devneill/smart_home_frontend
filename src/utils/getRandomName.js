export default function getRandomName() {
  const names = [
    "Enrique",
    "West",
    "Damien",
    "Braun",
    "Ellie",
    "Osborne",
    "Cierra",
    "Vega",
    "Alden",
    "Cantrell",
    "Kierra",
    "Gentry",
    "Pierre",
    "Cox",
    "Thomas",
    "Crane",
  ];
  const selector = Math.floor(Math.random() * names.length);
  return `${names[selector]}'s device`;
}
