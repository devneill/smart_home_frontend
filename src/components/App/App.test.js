import { cleanup, render, screen, waitFor } from "../../utils/customRender";
import userEvent from "@testing-library/user-event";
import { createMemoryHistory } from "history";
import React from "react";
import { Router } from "react-router-dom";
import { rest } from "msw";
import { setupServer } from "msw/node";
import "@testing-library/jest-dom/extend-expect";
import App from "../App";

const server = setupServer(
  rest.get(
    "https://canary-homework-test.herokuapp.com/devices/",
    (req, res, ctx) => {
      return res(ctx.json([{ id: 1, name: "Device 1" }]));
    }
  )
);

beforeAll(() => server.listen());
afterEach(() => server.resetHandlers(), cleanup);
afterAll(() => server.close());

it("should take a snapshot", () => {
  const { asFragment } = render(<App />);

  expect(asFragment(<App />)).toMatchSnapshot();
});

test("full app rendering/navigating", async () => {
  const history = createMemoryHistory();
  render(
    <Router history={history}>
      <App />
    </Router>
  );
  expect(screen.getByText(/all devices/i)).toBeInTheDocument();
  expect(screen.getByAltText(/loading-logo/i)).toBeInTheDocument();

  await waitFor(() => screen.getByText(/device 1/i));
  expect(screen.getByText(/device 1/i)).toBeInTheDocument();

  const leftClick = { button: 0 };
  userEvent.click(screen.getByText(/add device/i), leftClick);

  expect(screen.getByLabelText(/device name/i)).toBeInTheDocument();
  expect(screen.getByLabelText(/choose sensor type/i)).toBeInTheDocument();
});

xtest("landing on an invalid route", () => {
  const history = createMemoryHistory();
  history.push("/someinvalidroute");
  render(
    <Router history={history}>
      <App />
    </Router>
  );

  expect(screen.getByText(/page not found/i)).toBeInTheDocument();
});
