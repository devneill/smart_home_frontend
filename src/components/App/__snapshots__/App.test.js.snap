// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`should take a snapshot 1`] = `
<DocumentFragment>
  <header
    class="sc-jRQAMF cNPwjv"
  >
    <img
      alt="logo"
      class="sc-iCfLBT jTrKCv"
      src="logo.png"
    />
  </header>
  <div
    class="sc-hKwCoD bOrexo"
  >
    <div
      class="sc-eCImvq gJiXEx"
    >
      <div
        class="sc-furvIG cMtTkC"
      >
        <h1>
          All devices
        </h1>
        <button
          class="sc-pVTma fIxQiW"
        >
          Add device
        </button>
      </div>
      <img
        alt="loading-logo"
        class="loading-logo"
        src="logo.png"
      />
    </div>
  </div>
</DocumentFragment>
`;
